pkg/combination/combination.go

```diff
@@ -58,28 +75,69 @@ func WithSolveAhead() StreamOption {
 }
 
 // Next gets the next combination from the stream.
-// TODO: Currently this does not solve each combination iteratively and will
-//       need to do so in the future to ensure an optimal use of memory. This
-//       is currently more of a stub to allow consumer packages to maintain its
-// 		 interface.
+// By using this, the stream will solve each combination
+// iteratively.
 func (cs *stream) Next(ctx context.Context) (map[string]string, error) {
-	if cs.solveAhead && !cs.solved {
-		if err := cs.solve(); err != nil {
-			return nil, err
-		}
+	if cs.solved {
+		return nil, nil
 	}
+	// Check to see if anymore combinations exist
 
-	if len(cs.combinations) == 0 {
-		if cs.solved {
-			return nil, nil
-		}
-		return nil, ErrCombinationsNotSolved
+	// Edge case: No parameterListFromArgs
+	if len(cs.parameterListFromArgs) == 0 {
+		cs.solved = true
+		return nil, ErrNoArgsSet
 	}
 
-	next := cs.combinations[0]
-	cs.combinations = cs.combinations[1:]
+	// comboList is a variable that holds a list of combinations to be returned.
+	comboList := map[string]string{}
 
-	return next, nil
+	// Generate the list of combinations based off current positions
+	for x := 0; x < len(cs.parameterListFromArgs); x++ {
+		combo := cs.args[cs.parameterListFromArgs[x]][cs.positionsMapInArgs[x]]
+		key := cs.parameterListFromArgs[x]
+		comboList[key] = combo
+	}
+
+	// Iterates through position map in reverse
+	// looking for the first updatable value then breaks loop.
+	// Otherwise, resets values to zero, we know to update last parameter based off i.
+	var i int
+	for i = len(cs.positionsMapInArgs) - 1; i > cs.lastUpdatedParameter; i-- {
+		if cs.positionsMapInArgs[i]+1 < len(cs.args[cs.parameterListFromArgs[i]]) {
+			cs.positionsMapInArgs[i]++
+			break
+		} else {
+			cs.positionsMapInArgs[i] = 0
+		}
+	}
+
+	// Checks to see if we need to update lastParameter based
+	// off value of i.
+	if i == cs.lastUpdatedParameter {
+		// Checks to see if this is the last argument of the parameter.
+		// Then updates parameter, and checks to see if the combination is solved.
+		if cs.positionsMapInArgs[cs.lastUpdatedParameter]+1 == len(cs.args[cs.parameterListFromArgs[cs.lastUpdatedParameter]]) {
+			cs.positionsMapInArgs[cs.lastUpdatedParameter] = 0
+			cs.lastUpdatedParameter--
+		}
+		// If combination is not solved, we find the next lastUpdatedParameter,
+		// if lastUpdatedParameter has only 1 argument. Also, will mark as solved
+		// if reach end up parameters.
+		runner := true
+		for (!cs.solved) && (runner) && (cs.lastUpdatedParameter != -1) {
```

Comment:
nit: Are the parenthesis needed here or are they to add clarity?

Reply:
Just for clarity, but aren't needed